plotly/basedatatypes.py

```diff
@@ -0,0 +1,1783 @@
+            layout = Layout()  # type: Layout
+        else:
+            layout = self._layout_validator.validate_coerce(layout)
+
+        self._layout_obj = layout
+        self._layout = deepcopy(self._layout_obj._props)
+        self._layout_obj._parent = self
+        self._layout_defaults = {}
+
+        # Frames
+        # ------
+        from plotly.validators import FramesValidator
+        self._frames_validator = FramesValidator()
+
+        if frames:
+            self._frame_objs = self._frames_validator.validate_coerce(frames)
+        else:
+            self._frame_objs = ()
+
+        # Message States
+        # --------------
+        self._relayout_in_process = False
+        self._waiting_relayout_callbacks = []
+        self._last_relayout_msg_id = 0
+
+        self._restyle_in_process = False
+        self._waiting_restyle_callbacks = []
+        self._last_restyle_msg_id = 0
+
+        # View count
+        # ----------
+        self._view_count = 0
+
+        # Context manager
+        # ---------------
+        self._in_batch_mode = False
+        self._batch_style_commands = {}  # type: typ.Dict[int, typ.Dict[str, typ.Any]]
+        self._batch_layout_commands = {}  # type: typ.Dict[str, typ.Any]
+        self._animation_duration_validator = animation.DurationValidator()
+        self._animation_easing_validator = animation.EasingValidator()
+
+        # SVG
+        # ---
+        self._svg_requests = {}
+
+        # Logging
+        # -------
+        self._log_plotly_commands = False
+
+    # Magic Methods
+    # -------------
+    def __setitem__(self, prop, value):
+        if prop == 'data':
+            self.data = value
+        elif prop == 'layout':
+            self.layout = value
+        elif prop == 'frames':
+            self.frames = value
+        else:
+            raise KeyError(prop)
+
+    def __getitem__(self, prop):
+        if prop == 'data':
+            return self.data
+        elif prop == 'layout':
+            return self.layout
+        elif prop == 'frames':
+            return self.frames
+        else:
+            raise KeyError(prop)
+
+    def __contains__(self, prop):
+        return prop in ('data', 'layout', 'frames')
+
+    # Data
+    # ----
+    @property
+    def data(self) -> typ.Tuple['BaseTraceType']:
+        return self._data_objs
+
+    @data.setter
+    def data(self, new_data):
+
+        # Validate new_data
+        new_data = self._data_validator.validate_coerce(new_data)
+
+
+        orig_uids = [_trace['uid'] for _trace in self._data]
+        new_uids = [trace.uid for trace in new_data]
+
+        invalid_uids = set(new_uids).difference(set(orig_uids))
+        if invalid_uids:
+            raise ValueError(('The trace property of a figure may only be assigned to '
+                              'a permutation of a subset of itself\n'
+                              '    Invalid trace(s) with uid(s): {invalid_uids}').format(invalid_uids=invalid_uids))
```

Comment:
A couple things:
1) Your line has more than 79 characters. https://www.python.org/dev/peps/pep-0008/#id19
2) Write the ValueError message something like this:

```
raise ValueError(
    'The trace property of a figure may only be assigned to '
    'a permutation of a subset of itself \n Invalid trace(s) with uid(s): ' 
    '{invalid_uids}'.format(invalid_uids=invalid_uids)
)
```

it addresses 1) and it's more clean IMO

Reply:
Yeah, I started this as an independent project originally and chose to use a 119 character line width at the time. We can do a mass reformatting commit at some point (probably just run everything through [`yapf`](https://github.com/google/yapf) like I do in the codegen logic)